### ISSUE TITLE
Difference between emscripten build and other builds

### ISSUE THREAD

**Issue by chriseth** [ethereum/solidity#3660]
The emscripten binary in current develop and the other three binaries behave differently:

```
> diff emscripten.txt ubuntu-trusty.txt 
2110c2110,2111
< test_b9defe1f2a345cf72f670d21415acbf0d526573f47ba32198391e45045a850ca.sol: ERROR
---
> test_b9defe1f2a345cf72f670d21415acbf0d526573f47ba32198391e45045a850ca.sol:C 60606040523415600e57600080fd5b609e8061001c6000396000f300606060405260043610603f576000357c0100000000000000000000000000000000000000000000000000000000900463ffffffff16806326121ff0146044575b600080fd5b3415604e57600080fd5b6054606a565b6040518082815260200191505060405180910390f35b60005a9050905600a165627a7a72305820859264b76c162ca840d6b3ddbccf78cda7d3335914048836961419f64a0a08fa0029
> test_b9defe1f2a345cf72f670d21415acbf0d526573f47ba32198391e45045a850ca.sol:C {"compiler":{"version":"0.4.21-nightly.2018.3.6+commit.aebd9bc7"},"language":"Solidity","output":{"abi":[{"constant":true,"inputs":[],"name":"f","outputs":[{"name":"val","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"}],"devdoc":{"methods":{}},"userdoc":{"methods":{}}},"settings":{"compilationTarget":{"test_b9defe1f2a345cf72f670d21415acbf0d526573f47ba32198391e45045a850ca.sol":"C"},"evmVersion":"byzantium","libraries":{},"optimizer":{"enabled":false,"runs":200},"remappings":[]},"sources":{"test_b9defe1f2a345cf72f670d21415acbf0d526573f47ba32198391e45045a850ca.sol":{"keccak256":"0x349bcda9924fae97b20e95beb519500a79f1556d91c6c00d94daf75f55f7d036","urls":["bzzr://64c61823606cac501afef7089e4d9167dc8f2741d1bb84f1fcc83e6f9d06ec92"]}},"version":1}
2556c2557,2558
< test_e1a7d79c301bb09fd37893a99d4c4dd1bc04ce7172fd32014707277ca0459a2a.sol: ERROR
---
> test_e1a7d79c301bb09fd37893a99d4c4dd1bc04ce7172fd32014707277ca0459a2a.sol:test 60606040523415600e57600080fd5b609e8061001c6000396000f300606060405260043610603f576000357c0100000000000000000000000000000000000000000000000000000000900463ffffffff16806326121ff0146044575b600080fd5b3415604e57600080fd5b6054606a565b6040518082815260200191505060405180910390f35b60005a9050905600a165627a7a72305820ee9002fbb5514cb8a15e0fc55ca9f97119a5a16aa3e0b016b3418084c6aba9df0029
> test_e1a7d79c301bb09fd37893a99d4c4dd1bc04ce7172fd32014707277ca0459a2a.sol:test {"compiler":{"version":"0.4.21-nightly.2018.3.6+commit.aebd9bc7"},"language":"Solidity","output":{"abi":[{"constant":false,"inputs":[],"name":"f","outputs":[{"name":"val","type":"uint256"}],"payable":false,"stateMutability":"nonpayable","type":"function"}],"devdoc":{"methods":{}},"userdoc":{"methods":{}}},"settings":{"compilationTarget":{"test_e1a7d79c301bb09fd37893a99d4c4dd1bc04ce7172fd32014707277ca0459a2a.sol":"test"},"evmVersion":"byzantium","libraries":{},"optimizer":{"enabled":false,"runs":200},"remappings":[]},"sources":{"test_e1a7d79c301bb09fd37893a99d4c4dd1bc04ce7172fd32014707277ca0459a2a.sol":{"keccak256":"0x5f2e1b6c69d3d9ac348e2060735e7608f606f4dd5af3b3f7c84402231c66dbb9","urls":["bzzr://2311aa68e7e1b7b6eecc46505f52b1c73265ff62073704749f6b4000aa877203"]}},"version":1}
4943c4945,4946
< test_b9defe1f2a345cf72f670d21415acbf0d526573f47ba32198391e45045a850ca.sol: ERROR
---
> test_b9defe1f2a345cf72f670d21415acbf0d526573f47ba32198391e45045a850ca.sol:C 60606040523415600e57600080fd5b60998061001c6000396000f300606060405260043610603e5763ffffffff7c010000000000000000000000000000000000000000000000000000000060003504166326121ff081146043575b600080fd5b3415604d57600080fd5b60536065565b60405190815260200160405180910390f35b60005a9050905600a165627a7a723058208bef83b18617dde26c032637d2750a08f4cc4dbb8ce7602ccb93d8130561ce220029
> test_b9defe1f2a345cf72f670d21415acbf0d526573f47ba32198391e45045a850ca.sol:C {"compiler":{"version":"0.4.21-nightly.2018.3.6+commit.aebd9bc7"},"language":"Solidity","output":{"abi":[{"constant":true,"inputs":[],"name":"f","outputs":[{"name":"val","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"}],"devdoc":{"methods":{}},"userdoc":{"methods":{}}},"settings":{"compilationTarget":{"test_b9defe1f2a345cf72f670d21415acbf0d526573f47ba32198391e45045a850ca.sol":"C"},"evmVersion":"byzantium","libraries":{},"optimizer":{"enabled":true,"runs":200},"remappings":[]},"sources":{"test_b9defe1f2a345cf72f670d21415acbf0d526573f47ba32198391e45045a850ca.sol":{"keccak256":"0x349bcda9924fae97b20e95beb519500a79f1556d91c6c00d94daf75f55f7d036","urls":["bzzr://64c61823606cac501afef7089e4d9167dc8f2741d1bb84f1fcc83e6f9d06ec92"]}},"version":1}
5389c5392,5393
< test_e1a7d79c301bb09fd37893a99d4c4dd1bc04ce7172fd32014707277ca0459a2a.sol: ERROR
---
> test_e1a7d79c301bb09fd37893a99d4c4dd1bc04ce7172fd32014707277ca0459a2a.sol:test 60606040523415600e57600080fd5b60998061001c6000396000f300606060405260043610603e5763ffffffff7c010000000000000000000000000000000000000000000000000000000060003504166326121ff081146043575b600080fd5b3415604d57600080fd5b60536065565b60405190815260200160405180910390f35b60005a9050905600a165627a7a723058209b18d0bf137c030230055c9729d4489be9cbe307f27b2d22c3629a64155372e40029
> test_e1a7d79c301bb09fd37893a99d4c4dd1bc04ce7172fd32014707277ca0459a2a.sol:test {"compiler":{"version":"0.4.21-nightly.2018.3.6+commit.aebd9bc7"},"language":"Solidity","output":{"abi":[{"constant":false,"inputs":[],"name":"f","outputs":[{"name":"val","type":"uint256"}],"payable":false,"stateMutability":"nonpayable","type":"function"}],"devdoc":{"methods":{}},"userdoc":{"methods":{}}},"settings":{"compilationTarget":{"test_e1a7d79c301bb09fd37893a99d4c4dd1bc04ce7172fd32014707277ca0459a2a.sol":"test"},"evmVersion":"byzantium","libraries":{},"optimizer":{"enabled":true,"runs":200},"remappings":[]},"sources":{"test_e1a7d79c301bb09fd37893a99d4c4dd1bc04ce7172fd32014707277ca0459a2a.sol":{"keccak256":"0x5f2e1b6c69d3d9ac348e2060735e7608f606f4dd5af3b3f7c84402231c66dbb9","urls":["bzzr://2311aa68e7e1b7b6eecc46505f52b1c73265ff62073704749f6b4000aa877203"]}},"version":1}
```

**Comment by chriseth** [comment]
The difference seems to be related to the new globals:
```
> cat test_e1a7d79c301bb09fd37893a99d4c4dd1bc04ce7172fd32014707277ca0459a2a.sol
		contract test {
			function f() returns (uint256 val) {
				return gasleft();
			}
		}
> solc test_e1a7d79c301bb09fd37893a99d4c4dd1bc04ce7172fd32014707277ca0459a2a.sol
Warning: This is a pre-release compiler version, please do not use it in production.
test_e1a7d79c301bb09fd37893a99d4c4dd1bc04ce7172fd32014707277ca0459a2a.sol:2:4: Warning: No visibility specified. Defaulting to "public".
			function f() returns (uint256 val) {
   ^
Spanning multiple lines.
test_e1a7d79c301bb09fd37893a99d4c4dd1bc04ce7172fd32014707277ca0459a2a.sol:1:3: Warning: Source file does not specify required compiler version!
		contract test {
  ^
Spanning multiple lines.
test_e1a7d79c301bb09fd37893a99d4c4dd1bc04ce7172fd32014707277ca0459a2a.sol:3:12: Error: Undeclared identifier.
				return gasleft();
				       ^-----^

```

**Comment by chriseth** [comment]
Ok, the error from emscripten is:
```
Internal exception in StandardCompiler::compileInternal: /root/project/libsolidity/ast/Types.cpp(2360): Throw in function virtual string dev::solidity::FunctionType::richIdentifier() const
Dynamic exception type: N5boost16exception_detail10clone_implIN3dev8solidity21InternalCompilerErrorEEE
std::exception::what: Unknown function location.
[PN3dev11tag_commentE] = Unknown function location.
```

So this is probably related to the fact that emscripten just generates more data and thus touches more code.

**Comment by chriseth** [comment]
@ekpyron could you please fix this? It seems to be due to the fact that the new function location you introduced is not handled in all the places.

**Comment by chriseth** [comment]
Error location is `libsolidity/ast/Types.cpp(2360)`

**Comment by axic** [comment]
The new kind is missing (blockhash or gasleft), whichever was merged.

**Comment by axic** [comment]
Also to avoid this, please add a test for those, actually it would make sense to extend the identifier test case (https://github.com/ethereum/solidity/blob/develop/test/libsolidity/SolidityTypes.cpp#L106) we have to all possible cases.

**Comment by chriseth** [comment]
@axic adding such a test does not help in this case. If you miss it, you miss it. We should change all switch statements to not include a default case (where feasible) to get a warning at compile time.

**Comment by axic** [comment]
Fixed by #3663.
